### PR TITLE
fix: reject duplicate funding proofs in legacy and taproot swaps

### LINKED ISSUE #950: bug: Reject duplicate funding outpoints in Legacy and Taproot maker verification
## Description
A malicious taker can submit the same incoming funding output more than once when sending contract data to a maker.
Each entry is currently validated independently, but the maker does not ensure that the referenced outpoints are unique across the complete message. As a result, the same on-chain value can be counted multiple times.

This affects both protocols:
Legacy: duplicate entries in `ProofOfFunding.confirmed_funding_txes`
Taproot: duplicate transactions in `TaprootContractData.contract_txs`

The maker can:
- Accept the duplicated funding proof as valid.
- Create multiple incoming swap coins backed by the same output.
- Inflate the calculated total incoming amount.
- Construct outgoing funding transactions using its own wallet for value that was never provided by the taker.

A later failure does not safely mitigate the issue. During negative-control testing:
- Legacy accepted four funding entries, including one duplicate, inflated the incoming amount, and created outgoing transactions before later failing on a signature-count mismatch.
- Taproot accepted duplicated contract data and increased the incoming total from 500,000 to 681,604 sats. It attempted a fourth outgoing allocation and stopped only because the test maker had no remaining confirmed UTXO:

## Expected Behavior

The maker must reject the message before:
- Creating incoming swap coins
- Calculating outgoing amounts
- Reserving or spending wallet UTXOs
- Creating or broadcasting outgoing funding transactions

### PR DESCRIPTION
**Pull request by 0xEgao**
# Pull Request

## Description

Reject duplicate funding outpoints submitted by malicious takers in both Legacy and Taproot swaps.

Adds integration-only taker behavior and coverage verifying that makers reject duplicate proofs before creating or broadcasting outgoing funding transactions.

## Related Issue(s)
Fixes #950 

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor / performance improvement
- [ ] Documentation update only
- [ ] CI / Docker / Build changes
- [ ] Other (please describe):

## Protocol Version(s) Affected
- [ ] Legacy (ECDSA — `messages.rs`, `contract.rs`, `handlers.rs`)
- [ ] Taproot-Musig2 (`messages2.rs`, `contract2.rs`, `handlers2.rs`)
- [x] Both
- [ ] Neither (infrastructure / tooling only)

> **Note:** When modifying protocol logic, changes must be made to **both** versions unless the
> scope is explicitly version-specific.

## Affected Component(s)
- [ ] **makerd** (background daemon)
- [ ] **taker** (CLI client)
- [ ] **maker-cli** (command-line tool)
- [x] Core protocol / cryptography / library
- [ ] **watch_tower** (contract monitoring & recovery)
- [ ] Docker / deployment scripts
- [ ] Tests / test framework
- [ ] Documentation (`docs/`)
- [ ] Other (please specify):

### Code Quality
- [x] I ran `cargo +nightly fmt --all` and committed the result
- [x] I ran `cargo +stable clippy --all-features --lib --bins --tests -- -D warnings` with zero warnings
- [x] I ran `cargo +stable clippy --examples -- -D warnings` with zero warnings
- [x] I ran `RUSTDOCFLAGS="-D warnings" cargo +nightly doc --all-features --document-private-items --no-deps` with zero warnings
- [ ] Pre-commit git hook passes (`ln -s ../../git_hooks/pre-commit .git/hooks/pre-commit` if not already set)

### Testing
- [x] All unit tests pass (`cargo test`)
- [x] Integration tests pass (`cargo test --features integration-test`)
- [x] Changes were manually tested on **regtest**
- [x] End-to-end maker ↔ taker swap tested (where applicable)
- [x] Test-only code is gated behind `#[cfg(feature = "integration-test")]`

### Documentation
- [ ] Relevant files in the `docs/` folder were updated
- [x] Complex logic is commented

### Security & Privacy (Critical)
- [x] This change preserves **trustlessness** and **atomic swap guarantees**
- [x] No regression in **sybil resistance** or **fidelity bond** logic
- [x] Tor anonymity and P2P message flow were reviewed
- [x] No new attack vectors or trust assumptions introduced
- [x] Edge cases and error handling considered
- [x] ZMQ subscription integrity (raw tx/block feed) is unaffected
- [x] `integration-test` feature flag is not reachable in production code paths

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved Coinswap proof verification to reject duplicate funding outpoints during legacy proof-of-funding checks.
  - Added stricter validation to reject duplicate Taproot contract outpoints in Taproot contract data.
  - Ensures swaps fail early (before any funding-related broadcast) for both Legacy and Taproot.
- **Tests**
  - Added an integration regression test covering the duplicate-outpoint scenario across both protocol variants, asserting early rejection behavior.

### CODE REVIEW

File: src/taker/taproot_swap.rs [hunk 1]
```diff
@@ -256,6 +256,38 @@ impl Taker {
...
+                pubkeys,
+                timelock_scripts,
+                internal_keys,
+                tap_tweaks,
+                contract_txs,
+                amounts,
+            );
+            #[cfg(feature = "integration-test")]
+            if self.behavior == super::api::TakerBehavior::DuplicateFundingOutpoint {
+                if let Some(first) = contract_txs.first().cloned() {
```
Comment (mojoX911) [inline src/taker/taproot_swap.rs]: This can be combined into a single code block with one feature gate.

File: src/maker/taproot_verification.rs [hunk 2]
```diff
@@ -88,6 +90,14 @@ pub(crate) fn verify_taproot_contract_data(
         ));
     }
 
+    let mut seen_outpoints = HashSet::with_capacity(data.contract_txs.len());
+    for tx in &data.contract_txs {
+        let contract_outpoint = OutPoint::new(tx.compute_txid(), 0);
+        if !seen_outpoints.insert(contract_outpoint) {
+            return Err(MakerError::General("Duplicate Taproot contract outpoint"));
+        }
+    }
```
Comment (mojoX911) [inline src/maker/taproot_verification.rs:101]: Add a comment here and explain the attack. State why we need to hard-error here instead of just ignoring it.

Corollary: Do we really need a hard error? What if we just ignore the duplicate and carry on?

**Review by mojoX911 — CHANGES_REQUESTED** [review]

File: src/maker/taproot_verification.rs (hunk 2 above)
Comment (mojoX911) [reply src/maker/taproot_verification.rs:101]: Same applies for the legacy handler route.

Comment (0xEgao) [reply src/maker/taproot_verification.rs:101]: Isn't it's better to error out here, because for example if the taker sended N contracts, and the maker retained N-1, this will break protocol consistency for the future steps in that swap.

Also as a honest party, a taker should never send a duplicate entry, so i think it's better to treat it as a violation here and abort the swap.

Comment (mojoX911) [reply src/maker/taproot_verification.rs:101]: Okay agreed. This is a violation for sure. One thing to make sure of is that after this rejection, maker's swap liquidity should not be tied and revert back to the previous value. Can you assert that in the test? 

Also, the comment at the code site explaining the attack is still missing.

File: src/maker/taproot_verification.rs [hunk 3]
```diff
@@ -88,6 +90,16 @@ pub(crate) fn verify_taproot_contract_data(
         ));
     }
 
+    // Individually valid duplicate contracts would inflate total incoming value,
+    // causing the maker to allocate outgoing funds that the taker never provided.
```
Comment (0xEgao) [inline src/maker/taproot_verification.rs:94]: Comment is added to explain the attack
cc: @mojoX911

File: src/maker/api.rs [hunk 4]
```diff
@@ -1302,6 +1303,9 @@ impl MakerTrait for MakerServer {
 
         let min_reaction_time = MIN_CONTRACT_REACTION_TIME;
         let mut hashvalue: Option<crate::protocol::Hash160> = None;
+        // Each proof can be valid on its own, but repeating one outpoint makes the
+        // maker count the same incoming value twice and fund excess outgoing value.
```
Comment (0xEgao) [inline src/maker/api.rs:1307]: Same here

**Review by mojoX911 — APPROVED** [review]
Ack
